firmware-support/bittide-sys/src/time.rs

```diff
@@ -0,0 +1,365 @@
+        }
+    }
+}
+
+impl ops::AddAssign<Duration> for Duration {
+    fn add_assign(&mut self, rhs: Duration) {
+        self.micros += rhs.micros;
+    }
+}
+
+impl ops::Sub<Duration> for Duration {
+    type Output = Duration;
+
+    fn sub(self, rhs: Duration) -> Duration {
+        Duration {
+            micros: self
+                .micros
+                .checked_sub(rhs.micros)
+                .expect("overflow when subtracting durations"),
+        }
+    }
+}
+
+/// The Clock struct is a hardware abstraction for the timekeeping peripheral.
+/// It provides methods for waiting and updating time.
+/// Typical usage:
+/// ```
+/// let addr = 0x1000_0000 as *const u32;
+/// let clock = Clock::new(addr);
+/// let now = clock.elapsed();
+/// clock.wait(Duration::from_millis(1));
+/// let later = clock.elapsed();
+/// let duration = later - now;
+/// assert!(duration > Duration::from_millis(1));
+/// ```
+
+#[derive(Debug, Clone)]
+pub struct Clock {
+    freeze_count: *mut u32,
+    counter: *const u64,
+    frequency: *const u64,
+}
+
+impl Clock {
+    /// Create a new Clock instance.
+    ///
+    /// # Safety
+    ///
+    /// `addr` needs to point to a mapped memory address for a timer component.
+    pub fn new(addr: *const u32) -> Clock {
+        unsafe {
+            Clock {
+                freeze_count: addr.cast_mut(),
+                counter: addr.add(1).cast::<u64>(),
+                frequency: addr.add(3).cast::<u64>(),
+            }
+        }
+    }
+
+    /// Wait for a `Duration`.
+    pub fn wait(&self, duration: Duration) {
+        let mut now = self.elapsed();
+        let target = now + duration;
+        while target > now {
+            now = self.elapsed();
+        }
+    }
+
+    /// Wait until we have passed an `Instant`.
+    pub fn wait_until(&self, target: Instant) {
+        let mut now = self.elapsed();
+        while target > now {
+            now = self.elapsed();
+        }
+    }
+
+    /// Update the clock and return the current Instant.
+    pub fn elapsed(&self) -> Instant {
+        self.freeze();
+        Instant {
+            clock_cycles: self.get_counter(),
+            frequency: self.get_frequency(),
+        }
+    }
+
+    /// Freezes the time counter.
+    fn freeze(&self) {
+        unsafe {
+            self.freeze_count.write_volatile(0);
+        }
+    }
+
+    /// Retrieves the current value of the time counter.
+    ///
+    /// Returns:
+    /// - The current value of the time counter as a `u64`.
+    fn get_counter(&self) -> u64 {
+        unsafe { self.counter.read_volatile() }
+    }
+
+    /// Retrieves the frequency of the time counter.
+    ///
+    /// Returns:
+    /// - The frequency of the time counter as a `u64`.
+    pub fn get_frequency(&self) -> u64 {
+        unsafe { self.frequency.read_volatile() }
+    }
+
+    pub fn self_test(&self) -> [(&str, bool); 13] {
+        self_test::self_test(Clock::new(self.freeze_count as *const u32))
+    }
```

Comment:
I'm not sure if this should be a "member function" of this type or if it should be a free-standing function in the `self_test` module instead. If we add new tests this return type would change, which would be an API breaking change.

Maybe a `-> impl Iterator<Item = (&'static str, bool)>` is better?

Reply:
I'll make it free standing only